### ISSUE TITLE
Data parsed from JSON objects should not require a name

### ISSUE THREAD

**Issue by KensingtonTech** [freemarker/fmpp#17]
Parsing of a javascript object fails if a name isn't specified with it.

**config.fmpp:** (produces error)
```
sources:infile.orig.html
outputFile:outfile.html
data:{json(/root/fmpp_test_bug/esa_alert.json)}
```

**config.good.fmpp:** (this succeeds with name of vars and modified source file to use this)
```
sources:infile.mod.html
outputFile:outfile.html
data:{vars: json(/root/fmpp_test_bug/esa_alert.json)}
```

The error produced is:

```
Note: Using the config.fmpp in the working directory.
Failed!
Failed to apply the value of the "data" setting.

Caused by:
TDD error: Function doesn't evalute to a Map, but to fmpp.models.JSONObjectNode, so it can't be merged into the hash.
Error location: line 3, column 7 in /root/fmpp_test_bug/config.fmpp:
data:{json(/root/fmpp_test/esa_alert.json)}
```

This is using FMPP version 0.9.15.

I will send the assignee source files offline.

**Comment by citizentim** [comment]
The supporting files have been sent offline

**Comment by ddekany** [comment]
Ah, of course. The error message is right, though of course we humans know that a `JSONObjectNode` is a Map-ish thing, and it could be merged. I should add some interface that contains `toMap()`, and if TDD finds that, then instead of giving up with error could get the value as a `Map` and merge it. That's a new feature though.

**Comment by citizentim** [comment]
Hello Mr. Dékány,

I was wondering whether you've given any more thought to implementing this.  The documentation does say that it should work:

"Note: Above, the applicants: label is required because this file defines a JSON array, so we need a name with which we can refer to it from the templates. If a JSON file defines a JSON object instead (like { "applicants": [ ... ], "otherStuff": ... }), then, if you omit the label, the top-level name-value pairs from the JSON object will be added to the data-model directly."

Kind Regards

**Comment by ddekany** [comment]
I couldn't find the time for it yet...

**Comment by citizentim** [comment]
I've managed to fix this in a couple of hours.  No idea if there are any negative side-effects but it's working for me after a dist build.

Interpreter.java:
                    if (o1 instanceof Map) {
                        map.putAll((Map) o1);
                    } 
                    else if (o1 instanceof fmpp.models.JSONObjectNode) {
                        fmpp.models.JSONObjectNode newo1 = (fmpp.models.JSONObjectNode) o1;
                        map.putAll(newo1.toMap() );
                    }

JSONObjectNode.java:
public class JSONObjectNode extends JSONNode implements TemplateHashModelEx {
...
    public Map toMap() {
      return this.map;
    }

**Comment by AbnerZheng** [comment]
hi, citizentim,
I've the same problem as you, and found your patch to fix it. Have you meet any negative side-effects now?

**Comment by ddekany** [comment]
Committed fix for this, will be part of 0.9.16. (It's much more involved than the patch above, though for most practical cases that should work too.)

**Comment by ddekany** [comment]
Fix released with FMPP 0.9.16.
